pallets/author-filter/src/lib.rs

```diff
@@ -0,0 +1,151 @@
+// Copyright 2019-2020 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Small pallet responsible determining which accounts are eligible to author at the current
+//! block height.
+//!
+//! Currently this pallet is tightly coupled to our stake pallet, but this design
+//! should be generalized in the future.
+//!
+//! Using a randomness beacon supplied by the `Randomness` trait, this pallet takes the set of
+//! currently staked accounts from pallet stake, and filters them down to a pseudorandom subset.
+//! The current technique gives no preference to any particular author. In the future, we could
+//! disfavor authors who are authoring a disproportionate amount of the time in an attempt to
+//! "even the playing field".
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::pallet;
+
+pub use pallet::*;
+
+#[pallet]
+pub mod pallet {
+
+	use frame_support::pallet_prelude::*;
+	use frame_support::traits::Randomness;
+	use frame_support::traits::Vec;
+	use frame_system::pallet_prelude::*;
+	use sp_core::H256;
+	use sp_runtime::Percent;
+
+	/// The Author Filter pallet
+	#[pallet::pallet]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	/// Configuration trait of this pallet.
+	#[pallet::config]
+	pub trait Config:
+		frame_system::Config + stake::Config + cumulus_parachain_upgrade::Config
+	{
+		/// The overarching event type
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		/// Deterministic on-chain pseudo-randomness used to do the filtering
+		type RandomnessSource: Randomness<H256>;
+	}
+
+	// This code will be called by the author-inherent pallet to check whether the reported author
+	// of this block is eligible at this height. We calculate that result on demand and do not
```

Comment:
so every time the author inherent calls `set_author`, a potentially new random subset of validators is chosen as valid block authors? It seems like a lot of work to call every time. Also, could you just keep calling it until you're in the valid random subset because it's not set once but recalculated every time? 

Would it be better if we did this calculation once per round, whenever a new validator set is chosen and then store the chosen subset in storage?

Reply:
Never mind, I think I get it. It's choosing a pseudo-random subset at every block.